renga/cli/_graph.py

```diff
@@ -128,6 +131,76 @@ def _output_keys(self):
         """Return a list of the output keys."""
         return [n for n, d in self.G.out_degree() if d == 0]
 
+    def _need_update(self):
+        """Yield all files that need to be updated."""
+        visited = set()
+
+        for key in nx.topological_sort(self.G):
+            assert key not in visited
+            need_update = []
+            node = self.G.nodes[key]
+            latest = node.get('latest')
+            if latest:
+                need_update.append(key)
+
+            for parent, _ in self.G.in_edges(key):
+                assert parent in visited
+                need_update.extend(self.G.nodes[parent]['_need_update'])
+
+            if not latest and need_update:
+                need_update.append(key)
+
+            self.G.nodes[key]['_need_update'] = need_update
+            visited.add(key)
+
+    def build_status(self, revision='HEAD'):
+        """Return files from the revision groupped by their status."""
+        index = self.repo.git.index if revision == 'HEAD' \
+            else IndexFile.from_tree(repo.git, revision)
+
+        current_files = set()
+
+        for filepath, _ in index.entries.keys():
+            if not filepath.startswith('.renga') and \
+                    filepath not in {'.gitignore', }:
+                self.add_file(filepath, revision=revision)
+                current_files.add(filepath)
+
+        # Prepare status info for each file.
+        self._need_update()
+
+        graph_files = list(sorted(
```

Comment:
a few things here:

* do you need to turn `graph_files` into a list? it could be very long. 
* since you are doing a groupby in the next step, the sort here doesn't seem necessary either.